File: 2021_lua/day12.lua
-- Advent Of Code 2021
-- Copyright (C) 2024  Lukas Heindl
-- 
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
-- 
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
-- 
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

local _M = {}

local function parse_file(file)
	local caves = {}
	local function add_edge(s,e)
		if not caves[s] then caves[s] = s:match("[A-Z]+") and {big=true, conn={}} or {big=false, conn={}} end
		if not caves[e] then caves[e] = e:match("[A-Z]+") and {big=true, conn={}} or {big=false, conn={}} end
		table.insert(caves[s].conn, e)
		table.insert(caves[e].conn, s)
	end
	for line in file:lines() do
		add_edge(line:match("([a-zA-Z]+)%-([a-zA-Z]+)"))
	end
	return caves
end

local function print_path(path)
	io.write("{")
	for n,c in pairs(path) do
		if c > 0 then
			io.write(n, "->", c, " , ")
		end
	end
	io.write("}\n")
end

local function search(caves, can_visit)
	local paths = 0
	local function searchR(curr, dst, path)
		path[curr] = path[curr] and path[curr] + 1 or 1
		if curr == dst then
			paths = paths + 1
		else
			for _,n in ipairs(caves[curr].conn) do
				if can_visit(caves, path, n) then
					searchR(n, dst, path)
				end
			end
		end
		path[curr] = path[curr] - 1
	end
	searchR("start", "end", {})
	return paths
end

local can_visit = {
	part1 = function (caves, path, node)
		if caves[node].big then return true end
		return not path[node] or path[node] == 0
	end,
	part2 = function (caves, path, node)
		if caves[node].big then return true end
		local cnt = path

		if not cnt[node] or cnt[node] == 0 then return true end
		if node == "start" then return false end
		for k,v in pairs(cnt) do
			if v >= 2 and not caves[k].big then return false end
		end
		return true
	end
}

local function part(file, can_visit)
	local caves = parse_file(file)
	local paths = search(caves, can_visit)
	return paths
end

function _M.part1(file)
	return part(file, can_visit.part1)
end

function _M.part2(file)
	return part(file, can_visit.part2)
end

return _M

File: 2020_C/day-12.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <math.h>

#include "./read_file.h"

/*
Advent Of Code 2020
Copyright (C) 2024  Lukas Heindl

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// exmplanation how the direction is to be understood
/* enum direction{east=0, north=1, west=2, south=3}; */

int part1(struct string_content* input){
	int x = 0;
	int y = 0;
	int dir = 0;
	/* printf("Location: %d %d\n", x, y); */

	for(int i=0; i < input->length; i++){
		switch(*input->content[i]){
			case 'N': y += atoi(input->content[i]+1); break;
			case 'S': y -= atoi(input->content[i]+1); break;
			case 'E': x += atoi(input->content[i]+1); break;
			case 'W': x -= atoi(input->content[i]+1); break;
			case 'L': dir = ((dir + atoi(input->content[i]+1)/90) % 4 + 4) % 4; break;
			case 'R': dir = ((dir - atoi(input->content[i]+1)/90) % 4 + 4) % 4; break;
			case 'F':
					  switch(dir){
						  case 0: x += atoi(input->content[i]+1); break;
						  case 1: y += atoi(input->content[i]+1); break;
						  case 2: x -= atoi(input->content[i]+1); break;
						  case 3: y -= atoi(input->content[i]+1); break;
					  };
		};
		/* printf("%s\tLocation: %d %d \t Dir: %d\n", input->content[i], x, y, dir); */
	}
	return abs(x)+abs(y);
}

int part2(struct string_content* input){
	int x_ship = 0;
	int y_ship = 0;

	int x_way = 10;
	int y_way = 1;
	/* printf("Location Ship: %ld %ld\n", x_ship, y_ship); */
	/* printf("Location Way:  %ld %ld\n\n", x_way, y_way); */

	for(int i=0; i < input->length; i++){
		int num; // used as temporary storage
		switch(*input->content[i]){
			case 'N': y_way += atoi(input->content[i]+1); break;
			case 'S': y_way -= atoi(input->content[i]+1); break;
			case 'E': x_way += atoi(input->content[i]+1); break;
			case 'W': x_way -= atoi(input->content[i]+1); break;

			case 'L':
					switch(atoi(input->content[i]+1)/90){
						case 1:
							num = x_way;
							x_way = -y_way;
							y_way = num;
							break;
						case 2:
							x_way = -x_way;
							y_way = -y_way;
							break;
						case 3:
							num = x_way;
							x_way = y_way;
							y_way = -num;
							break;
						default:
							printf("Error2\n");
					};
					break;
			case 'R':
					switch(atoi(input->content[i]+1)/90){
						case 1:
							num = x_way;
							x_way = y_way;
							y_way = -num;
							break;
						case 2:
							x_way = -x_way;
							y_way = -y_way;
							break;
						case 3:
							num = x_way;
							x_way = -y_way;
							y_way = num;
							break;
						default:
							printf("Error1\n");
					};
					break;
			case 'F':
					num = atoi(input->content[i]+1);
					x_ship += x_way * num;
					y_ship += y_way * num;
					break;
		};
		/* printf("%s\tLocation Ship: %ld %ld\n", input->content[i], x_ship, y_ship); */
		/* printf("%s\tLocation Way:  %ld %ld\n\n", input->content[i], x_way, y_way); */
	}
	return abs(x_ship)+abs(y_ship);
}

int main() {
	char* type;
	(void)type;

	struct string_content *input;
	input = read_file("./problems/day-12.dat", type);
	printf("Part1:\nManhatten distance: %d\n", part1(input));
	printf("\nPart2:\nManhatten distance: %d\n", part2(input));
	free_file(input);
	return EXIT_SUCCESS;
}

// bad try calculating with sin/cos does not work
/* dist = sqrt(x_way*x_way + y_way*y_way); */
/* deg = asin(y_way/dist); */
/* printf("Dist: %f\tdeg vorher: %f\t", dist, deg); */
/* deg -= (double)atoi(input->content[i]+1) * M_PI/180; */
/* printf("deg nachher: %f\n", deg); */

/* y_way = lround(sin(deg)*dist); */
/* x_way = lround(cos(deg)*dist); */
